app/lib/search/updater.dart

```diff
@@ -4,92 +4,107 @@
-      versionKeys.add(p.latestVersionKey);
-      devVersions[p.name] = p.latestDevVersion;
-      count++;
-      if (limit != null && limit <= count) break;
+  Future updateIndex(Task task) async {
+    await _currentBatchRunning();
+    _batch.add(task);
+    _taskCount++;
+    if (_batch.length < 20) {
+      _batchUpdateTimer ??= new Timer(const Duration(seconds: 10), () {
+        _updateBatch();
+      });
+    } else {
+      await _updateBatch();
     }
-    _logger.info('Found ${versionKeys.length} packages to update.');
-    if (versionKeys.isEmpty) return;
+  }
 
-    for (List<Key> slices in sliceList(versionKeys, 20)) {
-      final Key firstKey = slices.first;
-      final String firstPkg = firstKey.parent.id;
-      final String firstVersion = firstKey.id;
-      _logger.info('Updating packages staring with $firstPkg $firstVersion');
-      final List<PackageVersion> versions = await dbService.lookup(slices);
-      final List<PackageDocument> documents = versions
-          .map((pv) => new PackageDocument(
-                url: _toUrl(pv.package),
-                package: pv.package,
-                version: pv.version,
-                devVersion: devVersions[pv.package],
-                detectedTypes: pv.detectedTypes,
-                description: pv.pubspec.description,
-                lastUpdated: pv.shortCreated,
-                readme: pv.readmeContent,
-                popularity: mockScores[pv.package] ?? 0.0,
-              ))
-          .toList();
-      await index.addAll(documents);
-      _logger.info('Updated ${slices.length} packages.');
-    }
+  Future _updateBatch() async {
+    _batchUpdateTimer?.cancel();
+    _batchUpdateTimer = null;
 
-    _logger.info('Calling index.merge() ...');
-    await index.merge();
-    _logger.info('index.merge() completed.');
+    await _currentBatchRunning();
+    if (_batch.isEmpty) return;
+
+    final Completer completer = new Completer();
+    _ongoingBatchUpdate = completer.future;
+
+    try {
+      final List<Task> tasks = new List.from(_batch);
+      _batch.clear();
+      _logger.info(
+          'Updating index with ${tasks.length} packages [example: ${tasks.first
+              .package}]');
+      final List<PackageDocument> docs = await searchBackend
+          .loadDocuments(tasks.map((t) => t.package).toList());
+      await packageIndex.addAll(docs);
+      final bool doMerge = _firstPackageListCount != null &&
+          _taskCount >= _firstPackageListCount;
+      if (doMerge) {
+        _logger.info('Merging index after $_taskCount updates.');
+        await packageIndex.merge();
+        _logger.info('Merge completed.');
+      }
+    } finally {
+      completer.complete();
+      _ongoingBatchUpdate = null;
+    }
   }
 
-  void startPolling() {
-    if (_timer == null) {
-      _timer = new Timer.periodic(new Duration(hours: 4), (_) {
-        update();
-      });
-      update();
+  Future _currentBatchRunning() async {
```

Comment:
You could simplify the code by getting rid of this function (see comment above).